hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/SigTransactionMetadataBuilder.java

```diff
@@ -13,77 +13,106 @@
 import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
 import com.hederahashgraph.api.proto.java.TransactionBody;
 import edu.umd.cs.findbugs.annotations.NonNull;
 import edu.umd.cs.findbugs.annotations.Nullable;
 import java.util.ArrayList;
-import java.util.Collections;
 import java.util.List;
 
 /**
- * Metadata collected when transactions are handled as part of "pre-handle" needed for signature
- * verification. This class may have subclasses in the future.
- *
- * <p>NOTE: This class shouldn't exist here, and is something of a puzzle. We cannot add it to SPI,
- * because it includes a dependency on AccountStore. But we also cannot put it in the app module,
- * because doing so would cause service modules to have a circular dependency on the app module.
- * Maybe we need some kind of base module from which services can extend and put it there?
+ * Builds {@link SigTransactionMetadata} by collecting needed information collected when
+ * transactions are handled as part of "pre-handle" needed for signature verification. This class
+ * will be subclassed For e.g., we need a {@link TransactionMetadata} with an inner {@link
+ * TransactionMetadata} for schedule transactions.
  */
-public class SigTransactionMetadata implements TransactionMetadata {
-    protected List<HederaKey> requiredKeys = new ArrayList<>();
+public class SigTransactionMetadataBuilder {
+    protected final List<HederaKey> requiredKeys = new ArrayList<>();
+    protected ResponseCodeEnum status = OK;
+
     protected TransactionBody txn;
-    protected AccountStore store;
+    protected final AccountKeyLookup keyLookup;
     protected AccountID payer;
 
-    protected ResponseCodeEnum status = OK;
+    public SigTransactionMetadataBuilder(final AccountKeyLookup keyLookup) {
+        this.keyLookup = keyLookup;
+    }
 
-    public SigTransactionMetadata(
-            final AccountStore store,
-            final TransactionBody txn,
-            final AccountID payer,
-            final List<HederaKey> otherKeys) {
-        this.store = store;
-        this.txn = txn;
-        this.payer = payer;
-        requiredKeys.addAll(otherKeys);
-        addPayerKey();
+    /**
+     * Sets status on {@link TransactionMetadata}. It will be {@link ResponseCodeEnum#OK} if there
+     * is no failure.
+     *
+     * @param status status to be set on {@link TransactionMetadata}
+     * @return builder object
+     */
+    public SigTransactionMetadataBuilder status(final ResponseCodeEnum status) {
+        this.status = status;
+        return this;
     }
 
-    public SigTransactionMetadata(
-            final AccountStore store, final TransactionBody txn, final AccountID payer) {
-        this(store, txn, payer, Collections.emptyList());
+    /**
+     * Set payer for the transaction
+     *
+     * @param payer payer for the transaction
+     * @return builder object
+     */
+    public SigTransactionMetadataBuilder payer(final AccountID payer) {
```

Comment:
```suggestion
    public SigTransactionMetadataBuilder payer(@NonNull final AccountID payer) {
```

Is there any legitimate case for supporting null?

Reply:
No , added NonNull. And checked `Objects.requireNonNull(..`